### PR TITLE
do not require the SMTPUTF8 extension when not needed

### LINKED ISSUE #145: remote smtp server does not implement smtputf8 extension, required by message
mox fails to send mails to some SMTP servers with the following error:
```
mox[6651]: l=error m="storing delivery error" pkg=queue deliveryerror="remote smtp server does not implement smtputf8 extension, required by message, transient, transient" cid=18e66f38ee7 from=robot@infra.example.com attempts=0 cid=18e66f38eec permanent=false code=0 secode=
mox[6651]: l=error m="temporary failure delivering from queue" err="remote smtp server does not implement smtputf8 extension, required by message, transient, transient" pkg=queue msgid=169 recipient=laurent@example.com backoff=7m28s nextattempt=2024-03-24T17:45:05+01:00 cid=18e66f38ee7 from=robot@infra.example.com attempts=0 cid=18e66f38eec permanent=false code=0 secode=
```

The email comes from a Prometheus alert.

Prometheus Alertmanager → mox → recipient SMTP server

mox requires the SMTPUTF8 extension on the recipient SMTP server although none of the email headers contain non-ascii characters. I've checked files in the `data/queue/a/` folder, none of those files contains non-ascii characters. The command `grep -r -P "[\x80-\xFF]" data/queue/` gives no results.

Prometheus uses the golang SMTP Client. This SMTP client uses the SMTPUTF8 extension when the SMTP server supports it (which is the case with mox) even when it's not needed (see: https://pkg.go.dev/net/smtp#Client.Mail).

It seems that mox logs somewhere that the email sent by Prometheus needs the SMTPUTF8 extension, and then requires that the recipient SMTP server also supports this extension. Not sure about that, is this assumption correct?

Could it be possible that mox actively checks if the SMTPUTF8 extension is really mandatory (not rely on the received email), and that it does not requires this extension when it's not needed?

### PR DESCRIPTION
**Pull request by lmeunier**
fix #145

### CODE REVIEW

**Comment by mjl-** [comment]
Thanks, I'll have a closer look tomorrow. First thoughts: Nice, with tests! I'm wondering if "header" has decoded mime q/b-word encoded text into utf-8, which could cause us to use smtputf8 when not needed (it is encoded as ascii).

File: smtpserver/server.go
Comment (mjl-) [inline smtpserver/server.go:1903]: i think we should do the smtputf8-check earlier. already in cmdData. just above the comment "Prepare "Received" header." (github PR review don't seem to let me make a comment there...). because we start composing headers based on c.smtputf8 from then on.

i think we should also keep a separate flag for whether we use smtputf8 for the message, vs what the incoming transaction has. perhaps we should add a flag "msgutf8" or "outsmtputf8" or something similar, and use that when composing headers and when adding to the queue. the smtputf8 flag is used when parsing commands. setting it to false after a the incoming connection has enabled it could lead to confusing/incorrect errors.

it would also be good to add a log line when we change from smtputf8 to non-utf8.

File: smtpserver/server.go
Comment (mjl-) [inline smtpserver/server.go:1914]: Localpart.String() marshals the Localpart string type to an encoded string. that won't change whether the string has non-ascii. so we can just check for characters with "isASCII(string(c.mailFrom.Localpart))".

File: smtpserver/server.go
Comment (mjl-) [inline smtpserver/server.go:1921]: i've read a bit more into rfc 653[012], which confirms that headers of multiparts can also have utf8, which is also a reason to use smtputf8. so this check is not exhaustive.

for example, see this message by arnt: https://github.com/arnt/eai-test-messages/blob/master/attachment
a smaller message like this would be a good test to add.

the easiest option is probably to just read all headers of all parts, and check for non-ascii. we would need a parsed message.Part. it is currently parsed in message.From() at the beginning of cmdSubmit, but we probably need to refactor a little bit to parse the part earlier, and perhaps just pass it to message.From (we don't want to parse messages multiple times).

something like this will do the check:

```
	hasNonASCII := func(r io.Reader) bool {
		br := bufio.NewReader(r)
		for {
			b, err := br.ReadByte()
			if err == io.EOF {
				break
			}
			xcheckf(err, "read header")
			if b > 0x7f {
				return true
			}
		}
		return false
	}

	var walk func(p *message.Part) bool
	walk = func(p *message.Part) bool {
		if hasNonASCII(p.HeaderReader()) {
			return true
		}
		for _, pp := range p.Parts {
			if walk(&pp) {
				return true
			}
		}
		return false
	}

	fmt.Println("message needs smtputf8:", walk(&part))
```

**Review by mjl- — CHANGES_REQUESTED** [review]
i went through rfc653[012] to page in the requirements. made some suggestions based on that.

File: smtpserver/server.go
Comment (lmeunier) [reply smtpserver/server.go]: Done.

About the separate `msgsmtputf8` flag, I must confess that I've replaced all occurrences of `c.smtputf8` found in server.go with the new flag without checking much (tests are still all ok, so we must be good...).

File: smtpserver/server.go
Comment (lmeunier) [reply smtpserver/server.go]: Done.

But I'm not sure the way I did it to avoid parsing the message twice is the best choice.

File: smtpserver/server_test.go
Comment (lmeunier) [inline smtpserver/server_test.go:1828]: Not sure about this one.

The utf8 character is only in the recipient domain, not the localpart. If I'm not wrong, the SMTPUTF8 extension is required **only** if the domain is not _translated_ to punycode. So this test should fail?

File: smtpserver/server.go
Comment (mjl-) [reply smtpserver/server.go]: yes, this is almost entirely fine. but the few calls to newParser should use c.smtputf8.

it's hard to trigger a problem around that: you would have to trigger an error in cmdData after setting msgsmtputf8 to false, then (as client) do another rcptto with a utf8 address. not even sure that would be a valid sequence of commands.

File: smtpserver/server.go
Comment (mjl-) [reply smtpserver/server.go]: looks good to me!

File: smtpserver/server_test.go
Comment (mjl-) [reply smtpserver/server_test.go:1828]: well, the smtpclient will send that literal recipient address to the smtpserver, with the utf8 extension, so smtputf8 and msgsmtputf8 will be true. the rcptto address is also included in the test message in the To header.
i think the test should be for a rcptto address with punycode a-labels (and also use a subdomain under example.org, not a domain that could actually exist, e.g. xn--vg8h.example.org for 🌏️.example.org).

File: smtpserver/server.go
Comment (mjl-) [inline smtpserver/server.go:1822]: style nit: the attributes shouldn't have spaces. it may even break the logfmt we are generating. we could just keep the names "smtputf8" and "msgsmtputf8". i also try to keep the messages lowercase, though that's not applied consistently.

File: smtpserver/server_test.go
Comment (mjl-) [inline smtpserver/server_test.go:1802]: nit: the parameter before clientSmtputf8 is eightbitmime, which we may want to set to true for future strictness checks.

**Review by mjl- — CHANGES_REQUESTED** [review]
Awesome, a few more comments, should be ready to merge next.

File: smtpserver/server.go
Comment (lmeunier) [reply smtpserver/server.go:1663]: I found two other functions similar to `hasNonASCII`:
* `isASCII` in message/compose.go:169
* `isASCII` in publicsuffix/list.go:162

Maybe we should refactor those functions and declare it only once ? (but where?)

File: smtpserver/server.go
Comment (lmeunier) [reply smtpserver/server.go]: As long as this PR is merged, I'll be happy. :)
Having a PR fully finished, or applying some style changes after merging, both are good for me. So, it's up to you to merge this PR when you decide it's content is worth merging.

File: smtpserver/server.go
Comment (mjl-) [reply smtpserver/server.go:1663]: we could merge it, but not sure if worth the trouble. indeed due to finding a place to put it (i think moxio would be a candidate, it's most like a "misc util" package. also, publicsuffix is what i consider a "leaf package", not great to include a dependency on a misc package. and then there is the required error handling that comes with reading from an io.Reader. since it's a small function, may be better to leave duplicates.

File: smtpserver/server.go
Comment (mjl-) [reply smtpserver/server.go]: i pushed a squashed merge commit just now. not sure if this was the best way to do it. i made a minor tweak afterwards to keep staticcheck happy.

thanks for the fix, much appreciated. looking forward to more. (:

Comment (lmeunier) [reply smtpserver/server.go]: About the squashed merge, I'm not sure either it's best way to do it. With github, the workflow is to merge this PR which will create a lot of commits in the main branch. To be honest, I'm not a big fan of the github PR workflow, I personally prefer the gerrit workflow (based on amended commit vs adding new commits) which gives a cleaner git history (only one commit per contribution).

But, anyway, thanks for merging this PR. :)

**Comment by lmeunier** [comment]
So, I guess I can close this PR now.
Closed by https://github.com/mjl-/mox/commit/9c5d2341629f598c298f4b9ae14009d108d814af

**Pull request closed without merging**
